mongorestore/restore.go

```diff
@@ -332,26 +332,87 @@ func (restore *MongoRestore) RestoreIntent(intent *intents.Intent) Result {
 	return result
 }
 
-func convertLegacyIndexes(indexes []IndexDocument) []IndexDocument {
-	for i, index := range indexes {
-		for j, elem := range index.Key {
-			switch elem.Value {
-			case true:
-				indexes[i].Key[j].Value = 1
-				log.Logvf(log.Always, "convertLegacyIndexes: converting the legacy vaule of '%s' from '%s' to '%s' in index '%s' on collection '%s'",
-					indexes[i].Key[j].Key, "true", "1", indexes[i].Options["name"], indexes[i].Options["ns"])
+func convertLegacyIndexes(indexes []IndexDocument) {
+	for _, index := range indexes {
+		convertLegacyIndexKeys(index)
+		convertLegacyIndexOptions(index)
+	}
+}
+
+func convertLegacyIndexKeys(index IndexDocument) {
+	var converted bool
+	originalJSONString := createExtJSONString(index.Key)
+	for j, elem := range index.Key {
+		switch v := elem.Value.(type) {
+		case int32, int64, float64:
+			// Only convert 0 value
+			if v == 0 {
+				index.Key[j].Value = 1
+				converted = true
+			}
+		case primitive.Decimal128:
+			if isZero(v) {
+				index.Key[j].Value = 1
+				converted = true
+			}
+		case string:
+			// Only convert an empty string
+			if v == "" {
+				index.Key[j].Value = 1
+				converted = true
 			}
+		default:
+			// Convert all types that aren't strings or numbers
+			index.Key[j].Value = 1
+			converted = true
 		}
+	}
+	if converted {
+		newJSONString := createExtJSONString(index.Key)
+		log.Logvf(log.Always, "convertLegacyIndexes: converted index values '%s' to '%s' on collection '%s'",
+			originalJSONString, newJSONString, index.Options["ns"])
+	}
+}
 
-		for key := range index.Options {
-			if _, ok := validIndexOptions[key]; !ok {
-				delete(indexes[i].Options, key)
-				log.Logvf(log.Always, "convertLegacyIndexes: removing invalid option '%s' from index '%s' on collection '%s",
-					key, indexes[i].Options["name"], indexes[i].Options["ns"])
-			}
+func isZero(num primitive.Decimal128) bool {
```

Comment:
I would avoid all this and just compare to `primitive.ParseDecimal128("0")`.  Odds are no-one has ever created an index with a decimal 128 anyway.  (I'm intentionally ignoring that Decimal 128 "0.00" is different than "0" or "-0", etc.)